pkg/api/customization/vsphere/handler.go

```diff
@@ -0,0 +1,187 @@
+package vsphere
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/norman/api/access"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/util"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/rancher/pkg/ref"
+	v1 "github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3/schema"
+	client "github.com/rancher/types/client/management/v3"
+	"github.com/rancher/types/config"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	UnknownFinder = iota
+	SoapFinder
+	ContentLibraryManager
+	TagsManager
+	CustomFieldsFinder
+)
+
+var fieldNames = map[string]int{
+	"data-centers":      SoapFinder,
+	"hosts":             SoapFinder,
+	"data-stores":       SoapFinder,
+	"folders":           SoapFinder,
+	"networks":          SoapFinder,
+	"virtual-machines":  SoapFinder,
+	"templates":         SoapFinder,
+	"clusters":          SoapFinder,
+	"resource-pools":    SoapFinder,
+	"content-libraries": ContentLibraryManager,
+	"library-templates": ContentLibraryManager,
+	"tags":              TagsManager,
+	"tag-categories":    TagsManager,
+	"fields":            CustomFieldsFinder,
+}
+
+var dataFields = map[string]string{
+	"username": "vmwarevspherecredentialConfig-username",
+	"password": "vmwarevspherecredentialConfig-password",
+	"host":     "vmwarevspherecredentialConfig-vcenter",
+	"port":     "vmwarevspherecredentialConfig-vcenterPort",
+}
+
+type handler struct {
+	schemas       *types.Schemas
+	secretsLister v1.SecretLister
+}
+
+func NewVsphereHandler(scaledContext *config.ScaledContext) http.Handler {
+	return &handler{
+		schemas:       scaledContext.Schemas,
+		secretsLister: scaledContext.Core.Secrets(namespace.GlobalNamespace).Controller().Lister(),
+	}
+}
+
+func (v *handler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	res.Header().Set("Content-Type", "application/json")
+	fieldName := mux.Vars(req)["field"]
+	dc := req.FormValue("dataCenter")
+
+	if fieldName == "" || !validFieldName(fieldName) {
+		util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: invalid field name", httperror.NotFound.Code))
+		return
+	}
+
+	cc, err := v.getCloudCredential(req)
+	if err != nil {
+		util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.InvalidBodyContent.Code, err.Error()))
```

Comment:
The error code should match the message you are returning. This is true for all the cases below